### ISSUE TITLE
serialPort.getInputStream().read() never returns -1 at end of data on Linux

### ISSUE THREAD

**Issue by mijak-extenda** [NeuronRobotics/nrjavaserial#27]
We are communicating with an external device connected to a serial port. The following code works flawlessly on Windows, but when running on Linux (Suse) it blocks indefinately.

Below is a simplified version of our code with a lot of log-statemens and stuff removed, also the actual handling of the data after the input-stream is closed.

We can see from our logs that on both Windows and Linux exactly the same bytes are read from the InputStream, and buffer.postion is incremented identically on both platforms.

After the final byte is read, on Windows the code moves on to evaluate the if-statement after the while loop (if (buffer.position() != bufPos)) and execute the handling of the received data. But on Linux after the final byte is read nothing more happens. I've added log-statements everywhere to try and see what actually happens, and my only conclusion is that on Linux in.read() blocks indefinately after the final byte is received and never returns -1.  Any idea how to remedy this?

```
public void serialEvent(SerialPortEvent e) {
  int data;
  synchronized(bufferLock) {
    try {
        InputStream in = serialPort.getInputStream();
        int bufPos = buffer.position();

        while ((data = in.read()) > -1) {
                byte b = (byte) data;

                if (buffer.remaining() == 0) {
                    log.warn("Out of buffer space. Doubling buffer size.");
                    int newSize = buffer.capacity() * 2;
                    ByteBuffer newBuffer = ByteBuffer.allocate(newSize);
                    buffer.flip();
                    newBuffer.put(buffer);
                    buffer = newBuffer;
                 }

                buffer.put(b);
        }

        if (buffer.position() != bufPos) {
            // We have added to the buffer
            // Handle the received data
        }
    } catch (IOException ex) {
        log.error("Error reading data.", ex);
        disconnect();
    } 
  }
```

**Comment by MrDOS** [comment]
Have you set a timeout? According to [the comments in the native code](https://github.com/NeuronRobotics/nrjavaserial/blob/master/src/main/c/src/SerialImp.c#L3423), the behaviour you're observing is intentional. Correct? That's much harder to say. The [interface documentation](http://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html) doesn't mention timeouts at all.

An appropriate course of action might be to see what the [Apache Commons Net `SocketInputStream`](https://commons.apache.org/proper/commons-net/apidocs/org/apache/commons/net/io/SocketInputStream.html) does in this situation: conceptually, the problem they're dealing with is very similar to ours.

Either way, there is a bug here as the behaviour is inconsistent between platforms (assuming you weren't setting a timeout on Windows and then not setting one on Linux).

**Comment by ssalonen** [comment]
@MrDOS  I have been developing modbus binding for openhab home automation sofware. 

Recently I got a report from a windows user that he gets read errors with the binding. It seems that at least with  the user's configuration, the read() return immediately with -1 as return value, even though receiveTimeout and receiveThreshold has been set. With one linux user the timeout seems to work as expected. Unlike the bug title here, we have managed to get EOF on linux by setting too short timeout value (case 3 as documented in [this post](https://community.openhab.org/t/connection-pooling-in-modbus-binding/5246/72?u=ssalonen))

See [this post](https://github.com/openhab/openhab/pull/4226#issuecomment-204046492) for details.

Any ideas what to try out or is there some obvious fix for this kind of behaviour? 
Is it normal to expect different behaviour different OSs? Or is this somehow related to the devices (e.g. USB/RS232 converter) used?
